pynsot/models.py

```diff
@@ -7,98 +7,550 @@
+    `raw` is provided
 
-    def __getitem__(self, val):
-        return self.__dict__[val]
+    __getitem__ and __iter__ are set so you can render what the NSoT payload
+    will look like by doing dict(obj) and fetch individual keys like a
+    dictionary with obj['attributes']
 
-    def __setitem__(self, key, val):
-        self.__dict__[key] = val
+    Attributes:
+        identifier (str)
+        rclient (pynsot.vendor.slumber.Resource)
+        rtype (str)
+        logger (logging.Logger)
+        payload (dict): This represents the exact payload sent to NSoT server
+        errors (list): Collection of any errors to happen during operations
+        last_error: The last error to happen
 
-    def __delitem__(self, key):
-        del self.__dict__[key]
+    Options:
+        site_id (int): (kwarg) Site ID for operations
+            Required unless `raw['site']`. If both provided, raw takes
+            precedence.
+        attributes (dict): (kwarg) Attributes to add to resource
+        client (pynsot.client.BaseClient): (kwarg) pynsot client
+            If client is not provided, `get_api_client` will be lazily loaded
+            on first API interaction. If you're dealing with many resource
+            objects that will talk to API, it'd be `n` less function calls to
+            pass in this parameter
+        raw (dict): (kwarg) Use this to pass in raw NSoT resource
+            Useful for instantiating objects as they return from API.
+    '''
+
+    __metaclass__ = ABCMeta
+
+    def __init__(
+        self,
+        site_id=None,
+        client=None,
+        raw={},
+        attributes={},
+        **kwargs
+    ):
+        self.logger = logging.getLogger(__name__)
+        self.errors = []
+        self.last_error = None
+        # Placeholder for .existing_resource() state
+        self._existing_resource = {}
+        self._payload = {}
+        # Parameter validations
+
+        # Site ID is required but can come from either kwarg or raw resource
+        # dict, latter being preferred.
+        if raw.get('site_id'):
+            site_id = raw['site_id']
+        elif site_id:
+            pass  # Already set
+        else:
+            raise TypeError('Resource requires site_id via param or `raw` key')
+
+        self.site_id = site_id
+        self.client = client
+        self.raw = raw
+        self.attributes = attributes
+
+        if not self.raw:
+            self.postinit(**kwargs)
+        else:
+            # This is done here because subclases do this in their postinit().
+            # If raw is passed, there's no reason to go through the postinit
+            # process which takes things like network_address, hostname, etc.
+            #
+            # Every subclasses init_payload method has a condition for 'if raw
+            # set raw' basically.
+            self.init_payload()
+
+    def postinit(self, **kwargs):
+        pass
+
+    def ensure_client(self):
```

Comment:
Please allow this to take `**kwargs**` and pass those onto `.get_api_client(**kwargs)`